### PR TITLE
gh-83395: Add curses.window.in_wch

### PR DESCRIPTION
**Pull request by asottile**
https://bugs.python.org/issue39214

* Issue: gh-83395

### CODE REVIEW

**Comment by GiorgosXou** [comment]
more than 2 years have passed, is this ever gonna be merged?

**Comment by krmaxson** [comment]
Providing this wide character method in the Python binding would let me ditch a soup of custom code. It is aggravating to have to duplicate the curses backing store to print non-ASCII! I've been staring at this pull request for two years -- I really hope it eventually gets merged. Thanks for tackling this!

**Comment by asottile** [comment]
I've rebased this against 3.12 (I originally targeted 3.9!)

**Comment by asottile** [comment]
> Providing this wide character method in the Python binding would let me ditch a soup of custom code. It is aggravating to have to duplicate the curses backing store to print non-ASCII! I've been staring at this pull request for two years -- I really hope it eventually gets merged. Thanks for tackling this!

fwiw, I've seen even outside reviews move things forward -- if you want to review the code

**Comment by GiorgosXou** [comment]
> I've been staring at this pull request for two years

@krmaxson me too, same thing, although at some point i got bored and I eventually diched the curses module and turned to the [`uni-curses`](https://github.com/unicurses/unicurses) module *(which I recently started maintaning and made [this](https://github.com/GiorgosXou/TUIFIManager) with it)*

**Comment by asottile** [comment]
aaaand rebased on 3.13

**Comment by asottile** [comment]
sure why not, let's do our yearly rebase -- now onto 3.14

File: Doc/library/curses.rst [hunk 1]
```diff
@@ -1017,6 +1017,18 @@ the following methods and attributes:
...
 
+   .. note::
+
+      ``inch`` only works for ASCII characters.  Use :meth:`in_wch` instead
+      for unicode support.
+
+.. method:: window.in_wch([x, y])
+
+   Return the wide character at the given position in the window.  The return
+   value is a 3-tuple ``(character, attributes, color_pair)``.
```
Comment (picnixz) [inline Doc/library/curses.rst:1028]:
```suggestion
   Return the wide character at the given position in the window as
   a triplet ``(character, attributes, color_pair)``.
```

File: Doc/library/curses.rst [hunk 2]
```diff
@@ -1017,6 +1017,18 @@ the following methods and attributes:
    Return the character at the given position in the window. The bottom 8 bits are
    the character proper, and upper bits are the attributes.
 
+   .. note::
+
+      ``inch`` only works for ASCII characters.  Use :meth:`in_wch` instead
+      for unicode support.
```
Comment (picnixz) [inline Doc/library/curses.rst:1023]:
```suggestion
   .. note::

   	  This method only works for ASCII characters. Use :meth:`in_wch`
   	  instead to retrieve the Unicode character at a given position.
```

File: Doc/library/curses.rst [hunk 3]
```diff
@@ -1017,6 +1017,18 @@ the following methods and attributes:
...
+
+.. method:: window.in_wch([x, y])
```
Comment (picnixz) [inline Doc/library/curses.rst:1025]:
```suggestion


.. method:: window.in_wch([x, y])
```

We separate methods by 2 blank lines (just check that my suggestion is correct before committing it).

Comment (picnixz) [inline Doc/library/curses.rst:1025]:
```suggestion

.. method:: window.in_wch([y, x])
```

File: Lib/test/test_curses.py [hunk 4]
```diff
@@ -1107,6 +1107,39 @@ def test_issue6243(self):
         curses.ungetch(1025)
         self.stdscr.getkey()
 
+    def test_in_wch(self):
+        stdscr = self.stdscr
+
+        if not hasattr(stdscr, 'in_wch'):
+            raise unittest.SkipTest('requires curses.window.in_wch')
```
Comment (picnixz) [inline Lib/test/test_curses.py:1114]: Do you need this kind of check or can you use `@requires_curses_window_meth('in_wch')`?

File: Lib/test/test_curses.py [hunk 5]
```diff
@@ -1107,6 +1107,39 @@ def test_issue6243(self):
...
+            color_attr = curses.color_pair(expected_pair)
+        else:
+            expected_pair = 0
+            color_attr = 0
+
+        def _norm(ch, attr, color):
+            # for some reason, curses returns some color bits here?
+            return ch, attr & ~curses.A_COLOR, color
+
+        for ch in ('a', '\xe9', '\u20ac', '\U0001f643'):
```
Comment (picnixz) [inline Lib/test/test_curses.py:1128]: I would use a `subTest` to locate failures if needed.

File: Lib/test/test_curses.py [hunk 6]
```diff
@@ -1107,6 +1107,39 @@ def test_issue6243(self):
...
+            expected = (ch, curses.A_BOLD, expected_pair)
+
+            stdscr.insstr(0, 0, ch, color_attr | curses.A_BOLD)
+            self.assertEqual(_norm(*stdscr.in_wch()), expected)
+
+            stdscr.addstr(0, 0, ch, color_attr | curses.A_BOLD)
+            self.assertEqual(_norm(*stdscr.in_wch(0, 0)), expected)
```
Comment (picnixz) [inline Lib/test/test_curses.py:1135]: Can we use something that is not at (0, 0) as well to check the ordering (y, x)?

File: Misc/NEWS.d/next/Library/2020-01-04-15-46-15.bpo-39214.z0398o.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+Add :func:`curses.window.in_wch` function - by Anthony Sottile.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2020-01-04-15-46-15.bpo-39214.z0398o.rst:1]:
```suggestion
Add :func:`curses.window.in_wch` function to get the Unicode character
at a given position. Patch by Anthony Sottile.
```

File: Modules/_cursesmodule.c [hunk 8]
```diff
@@ -1736,6 +1736,54 @@ _curses_window_inch_impl(PyCursesWindowObject *self, int group_right_1,
...
+#ifdef HAVE_NCURSESW
+/*[clinic input]
+_curses.window.in_wch
+
+    [
+    y: int
+        Starting Y-coordinate.
+    x: int
+        Starting X-coordinate.
+    ]
```
Comment (picnixz) [inline Modules/_cursesmodule.c:1748]:
```suggestion
    y: int
        Y-coordinate.
    x: int
        X-coordinate.
    ]
```

Let's use the same wording as for `inch`.

File: Modules/_cursesmodule.c [hunk 9]
```diff
@@ -1736,6 +1736,54 @@ _curses_window_inch_impl(PyCursesWindowObject *self, int group_right_1,
...
+    /
+
+Retrieve the wide character at the gven position in the window.
```
Comment (picnixz) [inline Modules/_cursesmodule.c:1751]:
```suggestion
Retrieve the Unicode character at the given position in the window.
```

File: Modules/_cursesmodule.c [hunk 10]
```diff
@@ -1736,6 +1736,54 @@ _curses_window_inch_impl(PyCursesWindowObject *self, int group_right_1,
...
+        ret = win_wch(self->win, &wcval);
+    }
+    if (PyCursesCheckERR(ret, "in_wch") == NULL) {
+        return NULL;
+    }
+
+    ret = getcchar(&wcval, wstr, &attr, &color_pair, NULL);
+    if (PyCursesCheckERR(ret, "getcchar") == NULL) {
+        return NULL;
+    }
```
Comment (picnixz) [inline Modules/_cursesmodule.c:1781]:
```suggestion
    if (ret == ERR) {
		PyErr_SetString(PyCursesError, "getcchar() returned ERR");
        return NULL;
    }
```

File: Modules/_cursesmodule.c [hunk 11]
```diff
@@ -1736,6 +1736,54 @@ _curses_window_inch_impl(PyCursesWindowObject *self, int group_right_1,
...
+    short color_pair;
+
+    if (group_right_1) {
+        ret = mvwin_wch(self->win, y, x, &wcval);
+    } else {
+        ret = win_wch(self->win, &wcval);
+    }
+    if (PyCursesCheckERR(ret, "in_wch") == NULL) {
+        return NULL;
+    }
```
Comment (picnixz) [inline Modules/_cursesmodule.c:1776]:
```suggestion
    if (group_right_1) {
        ret = mvwin_wch(self->win, y, x, &wcval);
      	if (ret == ERR) {
  		  	PyErr_SetString(PyCursesError, "mvwin_wch() returned ERR");
  		  	return NULL;
	  	}
    }
    else {
        ret = win_wch(self->win, &wcval);
        if (ret == ERR) {
  		  	PyErr_SetString(PyCursesError, "win_wch() returned ERR");
  		  	return NULL;
	  	}
    }
```

**Review by picnixz — COMMENTED** [review]
Some comments for this (I'm currently refactoring the curses module). By the way, is the term "color pair" documented on the page?

File: Modules/_cursesmodule.c [hunk 12]
```diff
@@ -1736,6 +1736,54 @@ _curses_window_inch_impl(PyCursesWindowObject *self, int group_right_1,
...
+
+The return value is a 3-tuple of (character, attributes, color_pair).
+[clinic start generated code]*/
+
+static PyObject *
+_curses_window_in_wch_impl(PyCursesWindowObject *self, int group_right_1,
+                           int y, int x)
+/*[clinic end generated code: output=846ca8a82f2ecab4 input=5c20d96b592b0e0b]*/
+{
+    int ret;
```
Comment (picnixz) [inline Modules/_cursesmodule.c:1761]: Ah maybe I should have said it in the other review, but I think we use `rtn` in general and not `ret`.

**Comment by asottile** [comment]
nevermind.  seriously not worth my time at this point

File: Modules/_cursesmodule.c (hunk 9 above)
Comment (picnixz) [reply Modules/_cursesmodule.c:1751]: Ah, wait, should we say wide character or Unicode character here (it's a `wchar_t` but we use it to indicate "Unicode" range so...)

**Review by picnixz — COMMENTED** [review]
There are also a bunch of tests that fail. I can't check them now but I'll have a look tomorrow.

**Pull request closed without merging**